packages/material-ui/src/Badge/Badge.d.ts

```diff
@@ -22,6 +47,4 @@ export type BadgeClassKey =
   | 'invisible'
   | 'dot';
 
-declare const Badge: React.ComponentType<BadgeProps>;
-
-export default Badge;
+export default function Badge(props: BadgeProps): JSX.Element;
```

Comment:
```suggestion
export default function Badge(props: BadgeProps): JSX.Element | null;
```
It's a bit overly cautious but some of our components do return `null`.

Reply:
`undefined` and `null` are both valid for a return type of `JSX.Element`

```typescript
function Badge(props: BadgeProps): JSX.Element {
  return null;
  return undefined;
  return <div></div>;
  // The rest are errors
  return 0;
  return '';
  return true;
}
```
The inferred type above is:
```typescript
function Badge(props: BadgeProps): true | JSX.Element | 0 | ""
```

Reply:
Is this some JSX domain specific weirdness or why would the above code be accepted? It fails with [similar code in the playground](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgEoUWAogGwgWwnGQG8BfAKFElkRQCkBlADVwKLGQgA9IQATAM5oMCbHkLESyShRgBXEGOAB7EMgBCcfgHMIACgCUALmRNWEjqQrJbyKBDDyo6kPJw4A3BTJA).

Reply:
This is not about whether this is a valid JSX constructor but that people don't assume we always return an element but might return `null`

Reply:
> Is this some JSX domain specific weirdness or why would the above code be accepted?

It checks for a valid JSX elements (which `null` is), when JSX is enabled it doesn't compare the actual definition, just what is a valid JSX element

> It fails with similar code in the playground.

There it checks the literal interface, not if it is a JSX element or not.

> This is not about whether this is a valid JSX constructor but that people don't assume we always return an element but might return null

My point is that `null` is a valid element, so adding `| null` is redundant

Reply:
> My point is that null is a valid element, so adding | null is redundant

It's redundant in a sense that both are valid JSX constructors. But that's just not the interface our components implement. 

I'm saying this a second time: I know what valid JSX constructors are. I know the caveats surrounding them in TypeScript compared to what is actually accepted in React.

I'm just saying that this is not the interface our components implement so please change it.

Reply:
> I'm just saying that this is not the interface our components implement so please change it.

I will, it requires changes upstream first, which is what I was trying to avoid :smile: